src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -782,18 +784,21 @@ void ShenandoahConcurrentGC::op_final_mark() {
         if (ShenandoahVerify) {
           heap->verifier()->verify_before_evacuation();
         }
-        // TODO: we do not need to run update-references following evacuation if collection_set->is_empty().
 
         heap->set_evacuation_in_progress(true);
```

Comment:
The new code at lines 796-801 makes sense. However, then why do we set evacuation in progress at line 788 above if there might be nothing to evacuate in the event that the collection set is empty, in which case the boolean parm should be like at line 799 below? If so, wouldn't the two conditions be synonymous? If not, what is the information in one that is not in the other. In that case, could the `set_evacuation_in_progress` method be renamed for greater clarity?

Reply:
I'll add a comment to clarify. The in-place promotions happen during the evacuation phase, _even when the collection set is empty_. We've discussed pulling this action into a separate/distinct phase, or _only_ allowing in-place-promotions  when the collection set is _not_ empty (thus eliminating the special case of cycles that _only_ perform in-place-promotions.